lib/urlwatch/reporters.py

```diff
@@ -134,7 +134,11 @@ def submit_all(cls, report, job_states, duration):
             if cfg['enabled']:
                 any_enabled = True
                 logger.info('Submitting with %s (%r)', name, subclass)
-                subclass(report, cfg, job_states, duration).submit()
+                if report.config['report'][subclass.__base_kind__].get('separate', False):
```

Comment:
I'm not necessarily opposed to going this route but this is the point where I would ask -- if we're going to allow `__base_kind__` to permeate the code, would it be better to enforce it's existence by throwing an error? (Perhaps with an abstract base class?) If we want to maintain backwards compatibility with unknown third party reporters, perhaps at least a warning that it ought to be defined and may be an error in a future release? Even if returning an empty dictionary works fine now, I don't think there's any way to guarantee it will continue working for all usages in the future and it strikes me as likely to lead to confusing behavior.

Reply:
The "cleanest" way would probably just to look at the MRO (method resolution order) for a class, this way finding the "base" kind automatically, and making it possible to get a "merged" dict of all config options (subclass configs overriding non-subclass configs).

The method resolution order can be found as [`__mro__`](https://docs.python.org/3/library/stdtypes.html#class.__mro__) on the class object.

Like this:

```
% PYTHONPATH=lib python3
Python 3.10.8 (main, Oct 13 2022, 09:48:40) [Clang 14.0.0 (clang-1400.0.29.102)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from urlwatch.reporters import ShellReporter
>>> ShellReporter.__mro__
(<class 'urlwatch.reporters.ShellReporter'>, <class 'urlwatch.reporters.TextReporter'>, <class 'urlwatch.reporters.ReporterBase'>, <class 'object'>)
>>> 
```

Then when using just `__kind__` on the base class:

```
diff --git a/lib/urlwatch/reporters.py b/lib/urlwatch/reporters.py
index 39150bc..46456f7 100644
--- a/lib/urlwatch/reporters.py
+++ b/lib/urlwatch/reporters.py
@@ -262,6 +262,8 @@ class HtmlReporter(ReporterBase):
 
 
 class TextReporter(ReporterBase):
+    __kind__ = 'text'
+
     def submit(self):
         cfg = self.report.config['report']['text']
         line_length = cfg['line_length']
```

You can get all "kinds" like this:

```
>>> from urlwatch.reporters import ShellReporter
>>> [getattr(cls, '__kind__', None) for cls in ShellReporter.__mro__]
['shell', 'text', None, None]
```

(note that the first `None` there is from `ReporterBase`, and the second `None` is from `object`)

Getting an "inherited config" could look like this:

```
>>> config = {}
>>> for key in reversed([getattr(cls, '__kind__', None) for cls in ShellReporter.__mro__]):
...     if key is not None:
...         config.update(self.report.config['report'][key])
... 
# here, you should be able to use "config"
```

I know this is a mouthful, but that's probably the nicest way to implement this.

Reply:
(in theory we could also give `ReporterBase` a `__kind__` of some kind (sic), e.g. `all` and then just make sure that the default config has `all: separate: false` by default, and then it would even magically work to:

1. Set `separate: true` for `all`, but not set it for anything else -> separate reports for everything
2. Set `separate: true` for `all`, and `separate: false` for `text` -> separate reports for all but text-based ones
3. Set `separate: false` for all`, `separate: true` for `text`, `separate: false` for `shell` -> separate reports only for text-based ones, except shell

Reply:
I think the `all` configuration idea is a good one but I think improving the "Reporter" section of the Configuration docs is a higher priority at the moment. The docs hint that there is a configuration hierarchy but you have to read `reporters.py` to find out what it is. I say "hint" because as an end user approaching the docs it isn't clear that the reporters are even python classes. Seems like at the minimum we could provide a hyperlink to `reporters.py` on github but better would be to just outline the hierarchy in the docs. Ideally the outline would be generated by an RST macro but that might not be worth the effort.

Reply:
Another thing I think would be really helpful in general is configuration validation. For example, I just found out the hard way that reporters' "enabled" key is required and if you omit it you get a KeyError traceback. Pydantic is my preferred solution for such problems because it encourages you to extract out any implementation details of the configuration and also makes documentation easier because all the information is in one place.

Reply:
Feel free to do the validation thing as a separate PR. This PR is nearly ready (see the comment I just posted), once that is fixed, I think this can be merged -- what do you think?

Reply:
Oh yeah, I wouldn't want to increase the scope of this PR.